# Reference-based Annotation

## Create the Reference
The first step in performing reference-based annotation is to select an 
annotated dataset to use as the reference. Here we will use one of the 
references created by the authors of SingleR and show how it can be
used with another tool.

Other reference datasets can be found in GEO or at a link provided by the 
authors of the reference dataset. However, to use a dataset as a reference
you will need both the scRNAseq data and the cell-type annotations. However,
GEO does not require authors to provide the cell-type annotations of their data,
so you may need to contact the authors directly to to get the annotations for 
some datasets.

```{r}
set.seed(9742)
# Download singleR reference data for immune cells
library(celldex)
ref <- DatabaseImmuneCellExpressionData()

```

Next we need to reformat the data to ensure it is compatible with the tool we
are using. We will be demonstrating scmap, which uses the SingleCellExperiment 
object, and assumes by default that gene names are found in a column named 
'feature_symbol' while the cell-type labels are in a column names 'cell_type1'.
In addition, scmap requires that you normalize and long-transform the reference data,
this has already been done for the SingleR reference data so we skip those steps here.

```{r}
colData(ref)$cell_type1 <- colData(ref)$label.fine
rowData(ref)$feature_symbol <- rownames(ref)

require(SingleCellExperiment)
ref_sce <- SingleCellExperiment(assays=list(logcounts=Matrix::Matrix(assays(ref)$logcounts)), 
			colData=colData(ref), rowData=rowData(ref))
```

Our reference data is ready to me used now. So lets process this data to build 
the index we will use to map our un-labelled data to. First, we select genes to use:

```{r}
require(scmap)
# Create scmap-cluster reference
ref_sce <- selectFeatures(ref_sce, suppress_plot=FALSE)


# Inspect genes selected by scmap
rownames(ref_sce)[which(rowData(ref_sce)$scmap_features)]
```

Now we can see the genes that scmap has chosen to use. If there are key markers missing
we can make sure they are included like this:

```{r}
my_key_markers = c("TRAC", "TRBC1", "TRBC2", "TRDC", "TRGC1", "TRGC2", "IGKC")
rowData(ref_sce)$scmap_features[rownames(ref_sce) %in% my_key_markers] <- TRUE
```

And we can remove genes that we think might be technical artefacts, such as mitochondria RNAs, like this:

```{r}
mt_genes <- rownames(ref_sce)[grep("^MT-", rownames(ref_sce))]
rowData(ref_sce)$scmap_features[rownames(ref_sce) %in% mt_genes] <- FALSE

scmap_feature_genes <- rownames(ref_sce)[which(rowData(ref_sce)$scmap_features)]
``

Now we build the reference profiles used in scmap-cluster. These profiles can be accessed and plotted
from inside the SingleCellExperiment object as follows:

```{r}
ref_sce <- scmap::indexCluster(ref_sce)
# Reference profiles - once generated original data is 
# not needed for scmap-cluster
heatmap(as.matrix(metadata(ref_sce)$scmap_cluster_index))
scmap_cluster_reference <- metadata(ref_sce)$scmap_cluster_index
```

From here on out scmap only needs this set of reference profiles, so if working with a very large reference 
one could save this index separately and load it directly to use when annotating new datasets. But since
that is not the case here, we will simply save this index to a variable for now. 

We will also demonstrate scmap-cell, so we will create that index as well. As before one would first
normalize and log-transform the reference data, and select genes to use. As we have already done that,
we need only run the command to build the scmap-cell index. There are two parameters we can set: M and k,
increasing M and k will give more accurate mapping but increase the size of the index, and the time needed
to map cells. Here we use the defaults:

```{r}
# Create scmap-cell reference
ref_sce <- scmap::indexCell(ref_sce)
scmap_cell_reference <- metadata(ref_sce)$scmap_cell_index
scmap_cell_metadata <- colData(ref_sce)
```

scmap-cell assigns cells in one dataset to their nearest neighbours in the reference dataset. One can use
any rule they like to transfer information, such as cell-type or pseudotime, from these nearest neighbours 
to the query data. Thus we need to store the associated metadata for the reference as well. Now we don't need 
to use our original reference dataset anymore.

## Assign cells to the reference.

Now we need to load our unlabeled dataset into R. Normal preprocessing including QC filtering, normalizing 
and log-transforming the data must be done prior to annotating. In addition, scmap is based on the 
SingleCellExperiment object, so if our data is stored as a Seurat object we must convert it to SingleCellExperiment
as shown below.

```{r}
require(Seurat)
data <- Read10X("10XPBMCs/pbmc3k_filtered_gene_bc_matrices/filtered_gene_bc_matrices/hg19/")
# Make SingleCellExperiment from raw matrix
query_sce <- SingleCellExperiment(assays=list(counts=data))

# Make SingleCellExperiment from Seurat
query_seur <- CreateSeuratObject(data)
query_sce <- as.SingleCellExperiment(query_seur)

# normalize the data using the scater package
query_sce <- scater::logNormCounts(query_sce)

# add feature_symbol column
rowData(query_sce)$feature_symbol <- rownames(query_sce)
```

Now you should have an entry in `assays(my_sce)` called `logcounts` with the log-normalized matrix. We are
now ready to annotate our data with scmap-cluster. Let's start with scmap-cluster:

```{r}
scmap_cluster_res <- scmapCluster(projection=query_sce, 
				index_list=list(immune1 = scmap_cluster_reference), 
				threshold=0.1)

# plot the results of our annotation
par(mar=c(10, 4, 0, 0))
barplot(table(scmap_cluster_res$combined_labs), las=2)

# add them to the query object
colData(query_sce)$scmap_cluster <- scmap_cluster_res$combined_labs

# Make a UMAP and add the cell-type annotations
query_sce <- scater::runUMAP(query_sce)
scater::plotReducedDim(query_sce, dimred="UMAP", colour_by="scmap_cluster")
```

Alternatively we could use scmap-cell, to find the 10 nearest neighbour then pick the annotation that
most common among those neighbour like this:

```{r}
nearest_neighbours <- scmapCell(projection=query_sce, 
	index_list = list(immune1 = scmap_cell_reference), 
	w=10)

#get metadata for the neighbours of each cell.
mode_label <- function(neighbours, metadata=scmap_cell_metadata$cell_type1) {
	freq <- table(metadata[neighbours])
	label <- names(freq)[which(freq == max(freq))]
	if (length(label) > 1) {return("ambiguous")}
	return(label)
}

scmap_cell_labs <- apply(nearest_neighbours$immune1$cells, 2, mode_label)

#add them to the query object
colData(query_sce)$scmap_cell <- scmap_cell_labs

# plot the results
par(mar=c(10, 4, 0, 0))
barplot(table(scmap_cell_labs), las=2)

# Make a UMAP and add the cell-type annotations
scater::plotReducedDim(query_sce, dimred="UMAP", colour_by="scmap_cell")
```

Another option compatible with the SingleCellExperiment Object is SingleR. As
before we need a reference and a query dataset. In the case of SingleR,
we need the entirety of the reference dataset, rather than generating a compressed
reference index as we did with scmap. In addition, running just this small example
demonstrates the difference in run time between the methods. 

```{r}
require(SingleR)
predictions <- SingleR(test=query_sce, ref=ref, labels=ref$label.fine)
colData(query_sce)$singleR <- predictions$pruned.labels

# plot the results
par(mar=c(10, 4, 0, 0))
barplot(table(scmap_cell_labs), las=2)

# Make a UMAP and add the cell-type annotations
scater::plotReducedDim(query_sce, dimred="UMAP", colour_by="singleR")
```

Another option is to integrate our query data with our reference data. Then
we simply transfer the labels from the annotated reference to the neighbouring 
query cells in the integrated dataset. Clustering the integrated data is
a common approach to transferring labels. We demonstrate how this could
be done with Harmony below. But the approach would be the same for any
integration tool.

Note: the SingleR reference is not single cells, but averages across many
cells. Thus we convert and downsample the reference for demonstration purposes.
For a real experiment, one would use the original single cells as the reference
when integrating datasets.


```{r}
set.seed(2891)
#Convert to Seurat Objects

#Add a count slot so we can convert to seurat objects
assays(ref_sce)[["counts"]] <- round(2^assays(ref_sce)[["logcounts"]]) -1
colnames(ref_sce) <- paste("cell", 1:ncol(ref_sce))

# Subset so both reference and query have same genes
ref_seur <- as.Seurat(ref_sce[rownames(ref_sce) %in% rownames(query_sce),])
ref_seur@active.ident <- factor(rep("reference", ncol(ref_seur)))

query_seur <- as.Seurat(query_sce[rownames(query_seur) %in% rownames(ref_sce),])
query_seur@active.ident <- factor(rep("query", ncol(query_seur)))

#downsample reference to be similar to query
totalUMI <- median(query_seur@meta.data$nCount_RNA)
ref_seur@assays$RNA@counts <- SampleUMI(ref_seur@assays$RNA@counts, max.umi=totalUMI, upsample=FALSE)

# Merge data
merged_seur <- merge(ref_seur, query_seur)
merged_seur@meta.data$source <- merged_seur@active.ident

# Scale & Integrate with harmony
require(harmony)
merged_seur <- NormalizeData(merged_seur)

# For consistency we use the scmap feature genes
VariableFeatures(merged_seur) <- scmap_feature_genes
#merged_seur <- FindVariableFeatures(merged_seur)

require(ggplot2)

merged_seur <- ScaleData(merged_seur)
merged_seur <- RunPCA(merged_seur)
merged_seur <- RunUMAP(merged_seur, dims=1:15)
DimPlot(merged_seur, reduction="umap")+ggtitle("Before Integration")

merged_seur <- RunHarmony(merged_seur, "source", dims.use=1:15)
merged_seur <- RunUMAP(merged_seur, dims=1:15, reduction="harmony")
DimPlot(merged_seur, reduction="umap")+ggtitle("After Integration")
```

Now that the data is integrated we will cluster the data and look at the annotations 
of the reference cells present in each cluster. As with all clustering, this may
require manual tuning of the resolution parameters to get the best labels.

```{r}
merged_seur <- FindNeighbors(merged_seur, reduction="harmony", dims=1:15)
merged_seur <- FindClusters(merged_seur, resolution=0.5)

DimPlot(merged_seur, reduction="umap")+ggtitle("After Integration")

table(merged_seur@meta.data$label.fine, 
		merged_seur@active.ident)
```


Here we have a table of the reference annotations (across rows) per cluster (across columns). 
We can manually label the clusters based on this table or we could create a rule to algorithmically label the 
clusters based on this table. Since there is only 10 clusters, we assign the labels manually.

```{r}
cluster_labs <- c("0"="ambiguous", 
	"1"="NK cells", 
	"2"="Monocytes, CD14+", 
	"3"="B cells, naive",
	"4"="T cells, CD4+, naive TREG",
	"5"="T cells, CD8+, naive",
	"6"="Monocytes, CD16+",
	"7"="T cells, CD4+, memory TREG ",
	"8"="T cells, CD4+, naive, stimulated ",
	"9"="T cells, CD8+, naive, stimulated")

#assign cluster label to the associated query cells
merged_seur@meta.data$annotation <- cluster_labs[merged_seur@meta.data$RNA_snn_res.0.5]

#add the results to the SingleCellExperiment Object and plot
query_sce$Harmony_lab <- merged_seur@meta.data$annotation[merged_seur@meta.data$source =="query"]
scater::plotReducedDim(query_sce, dimred="UMAP", colour_by="Harmony_lab")
```

# Refining / Consensus annotations

Once we have run several tools, we can use the consensus of the labels to get a 
more robust annotation. In this case we will simply use the most common label across tools
to assign the final automatically annotated label.

```{r}
annotation_columns <- c("scmap_cluster", "scmap_cell", "singleR", "Harmony_lab")

#Optional check how consistent the labelling was.
#head(colData(query_sce)[,annotation_columns])

get_consensus_label <- function(labels){
	labels <- labels[labels != "ambiguous"]
	if (length(labels) == 0) {return("ambiguous")}
	freq <- table(labels)
	label <- names(freq)[which(freq == max(freq))]
	if (length(label) > 1) {return("ambiguous")}
	return(label)
}

colData(query_sce)$consensus_lab <- apply(colData(query_sce)[,annotation_columns], 1, get_consensus_label)
scater::plotReducedDim(query_sce, dimred="UMAP", colour_by="consensus_lab")
```


